frame/bridge/relayer-game/src/lib.rs

```diff
@@ -0,0 +1,173 @@
+	type Event: From<Event<Self, I>> + Into<<Self as frame_system::Trait>::Event>;
+
+	// The currency use for bond
+	type RingCurrency: LockableCurrency<Self::AccountId, Moment = Self::BlockNumber>;
+
+	// A regulator to adjust relay args for a specific chain
+	type RelayerGameAdjustment: RelayerGameAdjustable;
+
+	// The target chain's relay module's API
+	type TargetChain: Relayable;
+}
+
+decl_event! {
+	pub enum Event<T, I: Instance = DefaultInstance>
+	where
+		AccountId = AccountId<T>,
+	{
+		/// TODO
+		TODO(AccountId),
+	}
+}
+
+decl_error! {
+	pub enum Error for Module<T: Trait<I>, I: Instance> {
+	}
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait<I>, I: Instance = DefaultInstance> as DarwiniaRelayerGame {
+		// Each `TcHeaderId` is a `Proposal`
+		// A `Proposal` can spawn many sub-proposals
+		pub Proposals
+			get(fn proposal)
+			: double_map
+				hasher(blake2_128_concat) TcBlockNumber<T, I>,
+				hasher(identity) TcHeaderHash<T, I>
+			=> Proposal<
+				AccountId<T>,
+				BlockNumber<T>,
+				RingBalance<T, I>,
+				TcBlockNumber<T, I>,
+				TcHeaderHash<T, I>
+			>;
+
+		// All the `TcHeader`s store here, **NON-DUPLICATIVE**
+		pub TcHeaders
+			get(fn tc_header)
+			: double_map
+				hasher(blake2_128_concat) TcBlockNumber<T, I>,
+				hasher(identity) TcHeaderHash<T, I>
+			=> RefTcHeader;
+
+		pub ChallengeTimes
+			get(fn challenge_time)
+			: double_map
+				hasher(blake2_128_concat) Round,
+				hasher(blake2_128_concat) TcBlockNumber<T, I>
+			=> T::BlockNumber;
+
+		// The finalize blocks' header's record id in darwinia
+		pub ConfirmedTcHeaderIds
+			get(fn confirmed_tc_header_id)
+			: TcHeaderId<TcBlockNumber<T, I>, TcHeaderHash<T, I>>;
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait<I>, I: Instance = DefaultInstance> for enum Call
+	where
+		origin: T::Origin
+	{
+		type Error = Error<T, I>;
+
+		fn deposit_event() = default;
+
+		#[weight = 0]
+		fn submit(origin, tc_header_thing: Vec<u8>) {
+			let relayer = ensure_signed(origin)?;
+			let tc_header_id = T::TargetChain::verify(&tc_header_thing)?;
+
+			// TcHeaders::insert(tc_header_id.0, tc_header_id.1, tc_header_thing);
+		}
+	}
+}
+
+#[derive(Clone, PartialEq, Encode, Decode, RuntimeDebug)]
+pub enum TcHeaderStatus {
+	// The header had been confirmed by game
+	Confirmed,
+	// The header had been confirmed by game but too old
+	// Means we might not use this header anymore so drop it to free the storage
+	Outdated,
+	// The header has not been judged yet
+	Unknown,
+	// The header is invalid
+	Invalid,
+}
+impl Default for TcHeaderStatus {
+	fn default() -> Self {
+		Self::Unknown
+	}
+}
+
+#[derive(Clone, Default, PartialEq, Encode, Decode, RuntimeDebug)]
+pub struct Proposal<AccountId, BlockNumber, Balance, TcBlockNumber, TcHeaderHash> {
+	// Current target chain's header id
+	id: TcHeaderId<TcBlockNumber, TcHeaderHash>,
+	// Will be confirmed automatically at this moment
+	confirm_at: BlockNumber,
```

Comment:
As discussed in last conversation, confirm_at should follow each rounds, not proposal.

Reply:
Yes, there's a new storage name `ChallengeTimes`, it'll trace all the round's challenge time.